File: py/portfolio.py
from bank import Bank
from money import Money
import functools
import operator


class Portfolio:
    def __init__(self) -> None:
        self.moneys = []

    def add(self, *moneys):
        self.moneys.extend(moneys)

    def evaluate(self, bank: Bank, currency: str) -> Money:
        total = 0.0
        failures = []
        for m in self.moneys:
            try:
                total += bank.convert(m, currency).amount
            except Exception as ex:
                failures.append(ex)

        if len(failures) == 0:
            return Money(total, currency)

        failureMessage = ",".join(f.args[0] for f in failures)
        raise Exception("Missing exchange rate(s):[" + failureMessage + "]")

File: js/portfolio.js
const Money = require("./money");

class Portfolio {
  constructor() {
    this.moneys = [];
  }

  add(...args) {
    this.moneys = [...this.moneys, ...args];
  }

  evaluate(bank, currency) {
    let failures = [];
    let total = this.moneys.reduce((res, el) => {
      let convertedMoney;
      try {
        convertedMoney = bank.convert(el, currency);
      } catch (error) {
        failures.push(error.message);
        return res;
      }
      return res + convertedMoney.amount;
    }, 0);
    if (failures.length === 0) {
      return new Money(total, currency);
    }
    throw new Error(`Missing exchange rate(s):[${failures.join(",")}]`);
  }
}
module.exports = Portfolio;
